bin/runtime-common/src/messages_xcm_extension.rs

```diff
@@ -0,0 +1,170 @@
+					e,
+					message.key.nonce
+				);
+				return MessageDispatchResult {
+					// TODO:check-parameter - setup uspent_weight?
+					unspent_weight: Weight::zero(),
+					dispatch_level_result: XcmBlobMessageDispatchResult::InvalidPayload,
+				}
+			},
+		};
+		let dispatch_level_result = match BlobDispatcher::dispatch_blob(payload) {
+			Ok(_) => {
+				log::debug!(
+					target: crate::LOG_TARGET_BRIDGE_DISPATCH,
+					"[XcmBlobMessageDispatch] DispatchBlob::dispatch_blob was ok - message_nonce: {:?}",
+					message.key.nonce
+				);
+				XcmBlobMessageDispatchResult::Dispatched
+			},
+			Err(e) => {
+				let e = match e {
+					DispatchBlobError::Unbridgable => "DispatchBlobError::Unbridgable",
+					DispatchBlobError::InvalidEncoding => "DispatchBlobError::InvalidEncoding",
+					DispatchBlobError::UnsupportedLocationVersion =>
+						"DispatchBlobError::UnsupportedLocationVersion",
+					DispatchBlobError::UnsupportedXcmVersion =>
+						"DispatchBlobError::UnsupportedXcmVersion",
+					DispatchBlobError::RoutingError => "DispatchBlobError::RoutingError",
+					DispatchBlobError::NonUniversalDestination =>
+						"DispatchBlobError::NonUniversalDestination",
+					DispatchBlobError::WrongGlobal => "DispatchBlobError::WrongGlobal",
+				};
+				log::error!(
+					target: crate::LOG_TARGET_BRIDGE_DISPATCH,
+					"[XcmBlobMessageDispatch] DispatchBlob::dispatch_blob failed, error: {:?} - message_nonce: {:?}",
+					e, message.key.nonce
+				);
+				XcmBlobMessageDispatchResult::NotDispatched(e)
+			},
+		};
+		MessageDispatchResult {
+			// TODO:check-parameter - setup uspent_weight?
+			unspent_weight: Weight::zero(),
+			dispatch_level_result,
+		}
+	}
+}
+
+/// [`XcmBlobHauler`] is responsible for sending messages to the bridge "point-to-point link" from
+/// one side, where on the other it can be dispatched by [`XcmBlobMessageDispatch`].
+pub trait XcmBlobHauler {
+	/// Runtime message sender adapter.
+	type MessageSender: MessagesBridge<Self::MessageSenderOrigin, XcmAsPlainPayload>;
+
+	/// Runtime message sender origin, which is used by MessageSender.
+	type MessageSenderOrigin;
+	/// Our location within the Consensus Universe.
+	fn message_sender_origin() -> Self::MessageSenderOrigin;
+
+	/// Return message lane (as "point-to-point link") used to deliver XCM messages.
+	fn xcm_lane() -> LaneId;
+}
+
+/// XCM bridge adapter which connects [`XcmBlobHauler`] with [`MessageSender`] and makes sure that
+/// XCM blob is sent to the [`pallet_bridge_messages`] queue to be relayed.
+pub struct XcmBlobHaulerAdapter<XcmBlobHauler>(sp_std::marker::PhantomData<XcmBlobHauler>);
+impl<HaulerOrigin, H: XcmBlobHauler<MessageSenderOrigin = HaulerOrigin>> HaulBlob
+	for XcmBlobHaulerAdapter<H>
+{
+	fn haul_blob(blob: sp_std::prelude::Vec<u8>) -> Result<(), HaulBlobError> {
+		let lane = H::xcm_lane();
+		let result = H::MessageSender::send_message(H::message_sender_origin(), lane, blob);
```

Comment:
I also wonder if XCM stuff cares about the weight of this `haul_blob` call? I mean - where from it is called? I assume that it is called when processing inbound XCMP/HRMP queue? Where from it is processed? In any way - if it is processed from `on_idle`, `on_finalize` or some similar things, then you shall know its weight there. Do you know how the weight is determined, @bkontur ?

Reply:
yes, it is triggered by `XcmpQueue->XcmExecutor::execute_xcm(instr::ExportMessage)`,

well, all fees/weights is still in TODOs https://github.com/paritytech/parity-bridges-common/issues/1677, and I would like to discuss that tomorrow

Reply:
lets hope, benchmark for `ExportMessage` should cover `haul_blob` weight: https://github.com/paritytech/parity-bridges-common/issues/1943

at least, I will write a test to BridgeHub runtime to trigger `XcmpQueue -> ExportMessage` to trigger some `max limit` for `MessageSender::send_message(`